### PR TITLE
Resolve empty relation opt for join types

### LINKED ISSUE #10967: Complete Possible Join Type Handling for EmptyRelation Propagation Rule
### Is your feature request related to a problem or challenge?

Currently on `main` only the `Inner` join type is considered when doing EmptyRelation propagation, but it's possible to infer if an EmptyRelation is propagatable for some other join types for some other empty conditions.

### Describe the solution you'd like

Fill out the join types per the TODO here, add or remove any that do or do not make sense: https://github.com/apache/datafusion/blob/a8847e1a825f8e588362430d99363672d8c3e7db/datafusion/optimizer/src/propagate_empty_relation.rs#L98-L109

- [x] For LeftOuter/LeftSemi/LeftAnti Join, only the left side is empty, the Join result is empty.
- [x] For LeftSemi Join, if the right side is empty, the Join result is empty.
- [x] For LeftAnti Join, if the right side is empty, the Join result is left side(should exclude null ??).
- [x] For RightOuter/RightSemi/RightAnti Join, only the right side is empty, the Join result is empty.
- [x] For RightSemi Join, if the left side is empty, the Join result is empty.
- [x] For RightAnti Join, if the left side is empty, the Join result is right side(should exclude null ??).
- [x] For Full Join, only both sides are empty, the Join result is empty.
- [ ] For LeftOut/Full Join, if the right side is empty, the Join can be eliminated with a Projection with left side columns + right side columns replaced with null values.
- [ ] For RightOut/Full Join, if the left side is empty, the Join can be eliminated with a Projection with right side columns + left side columns replaced with null values.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LorrensP-2158466**
## Which issue does this PR close?

Almost closes  #10967:

## Rationale for this change
Currently the PropagateEmptyRelation relation doesn't optimize all join types
This PR closes following cases
- For LeftAnti Join, if the right side is empty => the Join result is left side (including null)
- For RightAnti Join, if the left side is empty => the Join result is right side (including null)
- For Full Join, if both sides are empty => the Join result is empty

## What changes are included in this PR?
Added match statements that resolve the above cases.

## Are these changes tested?
Yes these changes are tested with unit tests and 3 added slt tests.

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 1]
```diff
@@ -142,13 +146,19 @@ impl OptimizerRule for PropagateEmptyRelation {
                             schema: join.schema.clone(),
                         }),
                     )),
+                    JoinType::LeftAnti if right_empty => {
```
Comment (alamb) [inline datafusion/optimizer/src/propagate_empty_relation.rs:149]: I double checked that an ANTI join (e.g. a `NOT IN`) is true when there are no tuples (aka the right side is emprt)

```sql
postgres=# select * from foo;
  x
-----
   1
   2
 NaN
(3 rows)

postgres=# select * from bar;
 y
---
(0 rows)

postgres=# select x NOT IN (SELECT y from bar) from foo;
 ?column?
----------
 t
 t
 t
(3 rows)
```

Thus I think this is a valid rewrite

**Review by alamb — APPROVED** [review]
Thank you @LorrensP-2158466 

cc @tshauck and @jonahgao

File: datafusion/optimizer/src/propagate_empty_relation.rs (hunk 1 above)
Comment (LorrensP-2158466) [reply datafusion/optimizer/src/propagate_empty_relation.rs:149]: Nice, thanks for the double check!

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 2]
```diff
@@ -142,13 +146,19 @@ impl OptimizerRule for PropagateEmptyRelation {
...
+                        Ok(Transformed::yes((*join.right).clone()))
+                    }
                     JoinType::RightAnti if right_empty => Ok(Transformed::yes(
                         LogicalPlan::EmptyRelation(EmptyRelation {
                             produce_one_row: false,
                             schema: join.schema.clone(),
                         }),
                     )),
-                    _ => Ok(Transformed::no(LogicalPlan::Join(join.clone()))),
+                    _ => Ok(Transformed::no(plan)),
```
Comment (jonahgao) [inline datafusion/optimizer/src/propagate_empty_relation.rs:161]: 👍

File: datafusion/optimizer/src/propagate_empty_relation.rs [hunk 3]
```diff
@@ -475,8 +485,39 @@ mod tests {
         assert_together_optimized_plan(plan, expected, eq)
     }
 
+    // TODO: fix this long name
+    fn assert_anti_join_empty_join_table_is_base_table(
+        anti_left_join: bool,
+    ) -> Result<()> {
+        // if we have an anti join with an empty join table, than the result is the base_table
```
Comment (jonahgao) [inline datafusion/optimizer/src/propagate_empty_relation.rs:492]:
```suggestion
        // if we have an anti join with an empty join table, then the result is the base_table
```

**Review by jonahgao — APPROVED** [review]
Looks good to me! Thanks @LorrensP-2158466

**Review by tshauck — APPROVED** [review]
Nothing to add over what's been said... LGTM!

**Comment by alamb** [comment]
I merged this PR up from main and plan to merge it when CI passes

**Comment by jonahgao** [comment]
Merged. Thanks @LorrensP-2158466 @alamb @tshauck

**Pull request merged**